File: services/flutter-app/code/packages/couchbase_lite_p2p/ios/Classes/MultipeerSyncManager.swift
import Foundation
import CouchbaseLiteSwift

class MultipeerSyncManager {
    private let database: Database
    private let peerGroupID = "com.example.counterapp"
    private let identityLabel = "com.example.counterapp.p2p.identity"
    
    var onStatusChanged: (([String: Any]) -> Void)?
    
    private var replicator: MultipeerReplicator?
    private var identity: TLSIdentity?
    
    private var isRunning = false
    private var connectedPeers: [PeerID] = []
    private var syncStatus = "Stopped"
    private var lastError: String?
    
    init(database: Database) {
        self.database = database
    }
    
    var currentStatus: [String: Any] {
        var peerIDStr = "unknown"
        if let pid = replicator?.peerID {
            peerIDStr = "\(pid)"
        }
        
        var status: [String: Any] = [
            "isRunning": isRunning,
            "myPeerID": peerIDStr,
            "connectedPeers": connectedPeers.count,
            "status": syncStatus
        ]
        
        if let error = lastError {
            status["error"] = error
        }
        
        return status
    }
    
    func start() {
        if isRunning { return }
        
        // Use a Task to handle async identity creation
        Task {
            do {
                // Add a small delay to ensure DB is fully ready and avoid race conditions
                try await Task.sleep(nanoseconds: 500_000_000) // 0.5 seconds

                // 1. Create or retrieve TLS identity
                identity = try createOrRetrieveIdentity()
                
                // 2. Create authenticator (accept all)
                let authenticator = MultipeerCertificateAuthenticator { _, _ in true }
                
                // 3. Create collection configurations
                let collection = try database.defaultCollection()
                let config = MultipeerCollectionConfiguration(collection: collection)
                
                // 4. Create MultipeerReplicator configuration
                let replConfig = MultipeerReplicatorConfiguration(
                    peerGroupID: peerGroupID,
                    identity: identity!,
                    authenticator: authenticator,
                    collections: [config]
                )
                
                // Attempt to force service type using KVC (desperate measure)
                // If this crashes, we know the property doesn't exist.
                // But if it works, it fixes the NoAuth error.
                // Note: serviceType usually doesn't include the underscore prefix in the setter
                // but Info.plist needs it.
                // Try "cbl-p2p" which matches _cbl-p2p._tcp
                // replConfig.setValue("cbl-p2p", forKey: "serviceType") 
                
                // 5. Create MultipeerReplicator
                replicator = try MultipeerReplicator(config: replConfig)
                
                // 6. Set up listeners
                setupEventListeners()
                
                // 7. Start
                replicator?.start()
                
                // 8. Update state
                isRunning = true
                syncStatus = "Running"
                notifyStatusChanged()
                
                print("✅ MultipeerReplicator started")
                
            } catch {
                print("❌ Failed to start MultipeerReplicator: \(error)")
                syncStatus = "Error: \(error.localizedDescription)"
                notifyStatusChanged()
            }
        }
    }
    
    func stop() {
        if !isRunning { return }
        
        replicator?.stop()
        replicator = nil
        
        isRunning = false
        syncStatus = "Stopped"
        connectedPeers.removeAll()
        notifyStatusChanged()
        
        print("🛑 MultipeerReplicator stopped")
    }
    
    private func createOrRetrieveIdentity() throws -> TLSIdentity {
        if let existing = try? TLSIdentity.identity(withLabel: identityLabel) {
            if existing.expiration > Date() {
                return existing
            }
            try? TLSIdentity.deleteIdentity(withLabel: identityLabel)
        }
        
        let attrs: [String: String] = [certAttrCommonName: "CounterApp"]
        let expiration = Calendar.current.date(byAdding: .year, value: 1, to: Date())!
        
        return try TLSIdentity.createIdentity(
            for: [.clientAuth, .serverAuth],
            attributes: attrs,
            expiration: expiration,
            label: identityLabel
        )
    }
    
    private func setupEventListeners() {
        guard let replicator = replicator else { return }
        
        // Status Listener
        _ = replicator.addStatusListener { [weak self] status in
            guard let self = self else { return }
            self.isRunning = status.active
            self.syncStatus = status.active ? "Active" : "Inactive"
            if let error = status.error {
                print("❌ Multipeer Replicator error: \(error.localizedDescription)")
                self.lastError = error.localizedDescription
            } else {
                self.lastError = nil
            }
            self.notifyStatusChanged()
        }
        
        // Peer Discovery Listener
        _ = replicator.addPeerDiscoveryStatusListener { [weak self] status in
            guard let self = self else { return }
            if status.online {
                if !self.connectedPeers.contains(status.peerID) {
                    self.connectedPeers.append(status.peerID)
                }
            } else {
                self.connectedPeers.removeAll { $0 == status.peerID }
            }
            self.notifyStatusChanged()
        }
    }
    
    private func notifyStatusChanged() {
        onStatusChanged?(currentStatus)
    }
}

File: services/flutter-app/code/packages/couchbase_lite_p2p/android/src/main/kotlin/com/example/couchbase_lite_p2p/couchbase_lite_p2p/MultipeerSyncManager.kt
package com.example.couchbase_lite_p2p.couchbase_lite_p2p

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import android.util.Log
import androidx.core.content.ContextCompat
import com.couchbase.lite.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.util.Calendar
import java.util.Date

class MultipeerSyncManager(
    private val context: Context,
    private val database: Database
) {
    
    private val peerGroupID = AppConfig.P2P_PEER_GROUP_ID
    private val identityLabel = AppConfig.P2P_IDENTITY_LABEL
    private val scope = CoroutineScope(Dispatchers.IO)
    
    companion object {
        private const val TAG = "MultipeerSync"
    }
    
    private var replicator: MultipeerReplicator? = null
    private var identity: TLSIdentity? = null
    
    data class P2PSyncState(
        val isRunning: Boolean = false,
        val myPeerID: String? = null,
        val connectedPeers: List<String> = emptyList(),
        val syncStatus: String = "Stopped",
        val error: String? = null
    )
    
    private val _syncState = MutableStateFlow(P2PSyncState())
    val syncState: StateFlow<P2PSyncState> = _syncState.asStateFlow()
    
    fun start() {
        scope.launch {
            try {
                if (_syncState.value.isRunning) return@launch
                
                if (!hasRequiredPermissions()) {
                    updateSyncState { it.copy(error = "Missing permissions") }
                    return@launch
                }
                
                // 1. Setup collections
                val collection = database.defaultCollection // Use default collection
                val collections = setOf(collection)
                
                // 2. Create or retrieve TLS identity
                identity = createOrRetrieveIdentity()
                
                // 3. Create authenticator
                val authenticator = MultipeerCertificateAuthenticator { _, _ -> true } // Accept all
                
                // 4. Create collection configurations
                val collectionConfigs = collections.map { 
                    MultipeerCollectionConfiguration.Builder(it).build() 
                }.toSet()
                
                // 5. Create MultipeerReplicator configuration
                val config = MultipeerReplicatorConfiguration.Builder()
                    .setPeerGroupID(peerGroupID)
                    .setIdentity(identity!!)
                    .setAuthenticator(authenticator)
                    .setCollections(collectionConfigs)
                    .build()
                
                // 6. Create MultipeerReplicator
                replicator = MultipeerReplicator(config)
                
                // 7. Set up listeners
                setupEventListeners()
                
                // 8. Start
                replicator?.start()
                
                val peerID = replicator?.peerId?.toString() ?: "unknown"
                updateSyncState { 
                    it.copy(isRunning = true, syncStatus = "Running", myPeerID = peerID, error = null) 
                }
                
                Log.i(TAG, "✅ MultipeerReplicator started with peer ID: $peerID")
                
            } catch (e: Exception) {
                Log.e(TAG, "❌ Failed to start MultipeerReplicator", e)
                updateSyncState { it.copy(isRunning = false, error = e.message) }
            }
        }
    }
    
    fun stop() {
        replicator?.stop()
        replicator = null
        updateSyncState { it.copy(isRunning = false, syncStatus = "Stopped", connectedPeers = emptyList()) }
    }
    
    private fun createOrRetrieveIdentity(): TLSIdentity {
        var identity = TLSIdentity.getIdentity(identityLabel)
        if (identity != null && identity.expiration.before(Date())) {
            TLSIdentity.deleteIdentity(identityLabel)
            identity = null
        }
        if (identity != null) return identity
        
        val certAttributes = mapOf(
            TLSIdentity.CERT_ATTRIBUTE_COMMON_NAME to "CounterApp",
            TLSIdentity.CERT_ATTRIBUTE_ORGANIZATION to "Couchbase Demo"
        )
        
        val calendar = Calendar.getInstance()
        calendar.add(Calendar.YEAR, 1)
        
        return TLSIdentity.createIdentity(
            setOf(KeyUsage.CLIENT_AUTH, KeyUsage.SERVER_AUTH),
            certAttributes,
            calendar.time,
            identityLabel
        )
    }
    
    private fun setupEventListeners() {
        val repl = replicator ?: return
        
        repl.addStatusListener { status ->
            scope.launch {
                updateSyncState { it.copy(isRunning = status.isActive) }
            }
        }
        
        repl.addPeerDiscoveryStatusListener { status ->
            scope.launch {
                val currentPeers = _syncState.value.connectedPeers.toMutableList()
                val peerIdStr = status.peer.toString()
                
                if (status.isOnline) {
                    if (!currentPeers.contains(peerIdStr)) currentPeers.add(peerIdStr)
                } else {
                    currentPeers.remove(peerIdStr)
                }
                
                updateSyncState { it.copy(connectedPeers = currentPeers) }
            }
        }
    }
    
    private fun updateSyncState(update: (P2PSyncState) -> P2PSyncState) {
        _syncState.value = update(_syncState.value)
    }
    
    private fun hasRequiredPermissions(): Boolean {
        val permissions = mutableListOf<String>()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            permissions.add(Manifest.permission.NEARBY_WIFI_DEVICES)
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            permissions.add(Manifest.permission.BLUETOOTH_ADVERTISE)
            permissions.add(Manifest.permission.BLUETOOTH_CONNECT)
            permissions.add(Manifest.permission.BLUETOOTH_SCAN)
            permissions.add(Manifest.permission.ACCESS_FINE_LOCATION)
        } else {
            permissions.add(Manifest.permission.ACCESS_FINE_LOCATION)
        }
        
        return permissions.all {
            ContextCompat.checkSelfPermission(context, it) == PackageManager.PERMISSION_GRANTED
        }
    }
}
